blues/reporters.py

```diff
@@ -0,0 +1,538 @@
+                atom=self.tables.Float32Atom(), shape=(0,))
+            self._handle.root.alchemicalLambda.attrs['units'] = str('dimensionless')
+
+        if set_protocolWork:
+            self._create_earray(where='/', name='protocolWork',
+                atom=self.tables.Float32Atom(), shape=(0,))
+            self._handle.root.protocolWork.attrs['units'] = str('kT')
+
+        if parameters:
+            if 'Logger' in self._parameters: self._parameters.pop('Logger')
+            paramjson = json.dumps(self._parameters)
+            self._encodeStringForPyTables(string=paramjson, name='parameters')
+
+        if set_environment:
+            try:
+                envout = subprocess.check_output('conda env export --no-builds', shell=True, stderr=subprocess.STDOUT)
+                envjson = json.dumps(yaml.load(envout), sort_keys=True, indent=2)
+                self._encodeStringForPyTables(envjson, name='environment')
+            except Exception as e:
+                print(e)
+                pass
+
+class BLUESHDF5Reporter(HDF5Reporter):
+    #This is a subclass of the HDF5 class from mdtraj that handles the reporting of
+    #the trajectory.
+    @property
+    def backend(self):
+        return BLUESHDF5TrajectoryFile
+
+    def __init__(self, file, reportInterval,
+                 title='NCMC Trajectory',
+                 coordinates=True, frame_indices=None,
+                 time=False, cell=True, temperature=False,
+                 potentialEnergy=False, kineticEnergy=False,
+                 velocities=False, atomSubset=None,
+                 protocolWork=True, alchemicalLambda=True,
+                 parameters=None, environment=True):
+
+        super(BLUESHDF5Reporter, self).__init__(file, reportInterval,
+            coordinates, time, cell, potentialEnergy, kineticEnergy,
+            temperature, velocities, atomSubset)
+        self._protocolWork = bool(protocolWork)
+        self._alchemicalLambda = bool(alchemicalLambda)
+        self._frame_indices = frame_indices
+        self._environment = bool(environment)
+        self._title = title
+        self._parameters = parameters
+
+    def report(self, simulation, state):
+        """Generate a report.
+        Parameters
+        ----------
+        simulation : simtk.openmm.app.Simulation
+            The Simulation to generate a report for
+        state : simtk.openmm.State
+            The current state of the simulation
+        """
+        if not self._is_intialized:
+            self._initialize(simulation)
+            self._is_intialized = True
+
+        self._checkForErrors(simulation, state)
+
+        args = ()
+        kwargs = {}
+        if self._coordinates:
+            coordinates = state.getPositions(asNumpy=True)[self._atomSlice]
+            coordinates = coordinates.value_in_unit(getattr(units, self._traj_file.distance_unit))
+            if self._frame_indices:
+                if simulation.currentStep not in self._frame_indices:
```

Comment:
I think you should turn self._frame_indices into a dictionary. The `if x in y` operation scales at worst O(n) for lists since it checks all the items in the list sequentially, while it scales O(1) for dictionaries, since it only has to check if the key is present in the dictionary. Usually it's a minor thing, but since we're doing this many times per iteration it could be worth optimizing.

Reply:
Can you elaborate how we could use a dictionary as opposed to a list? If we have to provide it a sequence of frames to store, it will have to be stored as a list anyways?

Reply:
It's only being called at the init so any slowdown would probably be negligible as opposed to the hundred of thousands of loop evaluations. It probably won't affect speed too much either way (unless you have a super long list), but if it's easy to implement I figure why not?